accepted/2020/mono-convergence/globalization-support-in-.net5.md

```diff
@@ -0,0 +1,82 @@
+
+Consult this [issue](https://github.com/android/ndk/issues/548#issuecomment-395561629) for detailed commentary on the challenges of making ICU an NDK API.
+
+### What Does This Mean For Us?
+The good news is that p-invoking from managed code works.  Assuming `libicuuc` continues to be white listed, this should not be a problem.  It would be wise if we explored using the `ICU4J`  API subset that’s exposed in the event they decide to tighten things up.  
+
+To summarize:
+
+* We can p-invoke into the system `libicuuc`.  Globalization for .net5 should work normally
+* Android 4.3 (API level 18) started bundling ICU 50 and our minimum version requirement is satisfied. 
+* OEM’s can potentially make life difficult by altering locales / globalization features
+* We should explore trying to interop with the `ICU4J` API and determine what the implications, if any are.
+
+### iOS/tvOS
+There is no build in support for ICU on apple mobile platforms. For our purposes of having highly compatible globalization mode across all platform we support we need to look into bringing ICU to ios/tvOS as well.
+
+System.Globalization.Invariant mode still need to work without any ICU dependency. How we are actually going to slice the actual data will be defined later based on the experience and existing ICU tooling.
+
+We are as expected not the first framework to hit this problem. Swift framework is dealing with same problems and they seem to prefer a solution which drop ICU dependency but that still does not look like final [decision](https://forums.swift.org/t/icu-usage-in-swift/20473).
+
+### WebAssembly
+The intention is set invariant globalization mode as the default mode for everyone for WebAssembly but at the same time allow a fallback to full globalization support when needed. This will come with additional cost but on the web platform we have little bit more flexibility how to deploy and share the data.
+
+ICU library consists from code and data. What we need to explore is if there is reasonable way to package the ICU data and store them in the cloud as a publicly available data blobs. This could give us flexibility to support full range of ICU data on demand if sliced into small data sets. We’ll still need to ship icu implementation but that could be also bound dynamically using dynamic linking.
+
+## ICU Packaging
+Due to lack of ICU on all platform we intend to support for .NET5 we need to look into how to package it for following RIDs
+
+* ios-x64
+* ios-arm
+* ios-arm64
+* tvos-x64
+* tvos-arm64
+* webassembly (browser mode)
+
+A default build of ICU normally results in over 16 MB of data, and a substantial amount of object code.
+
+The packages have to be split to implementation and data part and ideally the data parts will be also sliced in the way that the developers can select only relevant data for their app if they for example don’t target Chinese market.
```

Comment:
Would it make sense to statically link the ICU and the System.Globalization shim together, so that we only carry implementation for the the ICU APIs that the shim needs?

Reply:
Would that break app local scenarios? This way if people want to upgrade to a newer ICU, they couldn't because we carry the implementations that we built against, no?

Reply:
This paragraph is proposing to have our own custom build of ICU. There would not be a newer version unless we build it.

Reply:
Ah I see. I thought you were proposing that for all scenarios.

Reply:
I don't know what the implications are, if any, behind that.  I'm sure it's something we could determine during when implementing.